src/visualizer.h

```diff
@@ -0,0 +1,498 @@
+
+		{ql::__display__, { black, {
+			{}}}},
+		{ql::__display_binary__, { black, {
+			{}}}},
+		{ql::__nop_gate__, { black, {
+			{}}}},
+		{ql::__dummy_gate__, { black, {
+			{}}}},
+
+		{ql::__swap_gate__, { black, {
+			{ CROSS, 6, "", 0, black, black, black },
+			{ CROSS, 6, "", 0, black, black, black }}}},
+
+		{ql::__wait_gate__, { black, {
+			{}}}},
+		{ql::__classical_gate__, { black, {
+			{}}}}
+	};
+};
+
+struct CircuitData
+{
+	const unsigned int amountOfQubits;
+	const unsigned int amountOfClassicalBits;
+	const unsigned int amountOfCycles;
+};
+
+void visualize(const std::vector<ql::gate*> gates, const Layout layout);
+
+void validateLayout(const Layout layout);
+
+unsigned int calculateAmountOfBits(const std::vector<ql::gate*> gates, const std::vector<size_t> ql::gate::* operandType);
+
+void drawCycleNumbers(cimg_library::CImg<unsigned char>& image, const Layout layout, const CircuitData circuitData);
+void drawBitLine(cimg_library::CImg<unsigned char>& image, const Layout layout, const BitType bitType, const unsigned int row, const CircuitData circuitData);
+void drawGroupedClassicalBitLine(cimg_library::CImg<unsigned char>& image, const Layout layout, const CircuitData circuitData);
+void drawGate(cimg_library::CImg<unsigned char>& image, const Layout layout, const CircuitData circuitData, ql::gate* const gate);
+
+void drawGateNode(cimg_library::CImg<unsigned char>& image, const Layout layout, const CircuitData circuitData, const Node node, const NodePositionData positionData);
+void drawControlNode(cimg_library::CImg<unsigned char>& image, const Layout layout, const CircuitData circuitData, const Node node, const NodePositionData positionData);
+void drawNotNode(cimg_library::CImg<unsigned char>& image, const Layout layout, const CircuitData circuitData, const Node node, const NodePositionData positionData);
+void drawCrossNode(cimg_library::CImg<unsigned char>& image, const Layout layout, const CircuitData circuitData, const Node node, const NodePositionData positionData);
+
+//TODO: remove these functions for release version
+std::vector<ql::gate*> getTeleportationCircuit()
+{
+	std::vector<ql::gate*> gates;
+
+	ql::gate* gate1 = new ql::hadamard(1);
+	gate1->cycle = 0;
+	gates.push_back(gate1);
+
+	ql::gate* gate2 = new ql::cnot(1, 2);
+	gate2->cycle = 1;
+	gates.push_back(gate2);
+
+	ql::gate* gate3 = new ql::cnot(0, 1);
+	gate3->cycle = 3;
+	gates.push_back(gate3);
+
+	ql::gate* gate4 = new ql::hadamard(0);
+	gate4->cycle = 5;
+	gates.push_back(gate4);
+
+	ql::gate* gate5 = new ql::measure(1, 0);
+	gate5->cycle = 5;
+	gates.push_back(gate5);
+
+	ql::gate* gate6 = new ql::measure(0, 1);
+	gate6->cycle = 6;
+	gates.push_back(gate6);
+
+	ql::gate* gate7 = new ql::cnot(1, 2);
+	gate7->cycle = 7;
+	gates.push_back(gate7);
+
+	ql::gate* gate8 = new ql::cphase(0, 2);
+	gate8->cycle = 9;
+	gates.push_back(gate8);
+
+	return gates;
+}
+
+std::vector<ql::gate*> getDemoCircuit1()
+{
+	std::vector<ql::gate*> gates;
```

Comment:
move these test/example cases to test_visualizer.py unit test and code the example in OpenQL and call the visualizer as pass